cpp/include/raft/neighbors/brute_force_batch_k_query.cuh

```diff
@@ -0,0 +1,240 @@
+#include <raft/distance/distance_types.hpp>
+#include <raft/linalg/norm.cuh>
+#include <raft/matrix/slice.cuh>
+#include <raft/neighbors/brute_force.cuh>
+
+#include <raft/core/logger.hpp>
+
+namespace raft::neighbors::brute_force {
+/**
+ * @addtogroup brute_force
+ * @{
+ */
+
+/**
+ * @brief Brute force query over batches of k
+ *
+ * This class lets you query for batches of k. For example, you can get
+ * the first 100 neighbors, then the next 100 neighbors etc.
+ *
+ * Example usage:
+ * @code{.cpp}
+ * // create a brute_force knn index
+ * raft::device_resources res;
+ * auto index = raft::neighbors::brute_force::build(res,
+ *                                                  raft::make_const_mdspan(dataset.view()));
+ *
+ * // search the index in batches of 128 nearest neighbors
+ * auto search = raft::make_const_mdspan(dataset.view());
+ * for (auto & batch: batch_k_query(res, index, query, 128)) {
+ *  // batch.indices() and batch.distances() contain the information on the current batch
+ * }
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices in the source dataset
+ */
+template <typename T, typename IdxT = int64_t>
+class batch_k_query {
+ public:
+  /**
+   * @brief Construct a brute force batch k query object
+   *
+   * Constructs the batch_k_query - which lets you iterate over batches of
+   * nearest neighbors.
+   *
+   * @param[in] res
+   * @param[in] index The index to query
+   * @param[in] query A device matrix view to query for [n_queries, index->dim()]
+   * @param[in] batch_size The size of each batch
+   */
+  batch_k_query(const raft::resources& res,
+                const raft::neighbors::brute_force::index<T>& index,
+                raft::device_matrix_view<const T, int64_t, row_major> query,
+                int64_t batch_size)
+    : res(res), index(index), query(query), batch_size(batch_size)
+  {
+    auto metric = index.metric();
+
+    // precompute query norms, and re-use across batches
+    if (metric == raft::distance::DistanceType::L2Expanded ||
+        metric == raft::distance::DistanceType::L2SqrtExpanded ||
+        metric == raft::distance::DistanceType::CosineExpanded) {
+      query_norms = make_device_vector<T, int64_t>(res, query.extent(0));
+
+      if (metric == raft::distance::DistanceType::CosineExpanded) {
+        raft::linalg::norm(res,
+                           query,
+                           query_norms->view(),
+                           raft::linalg::NormType::L2Norm,
+                           raft::linalg::Apply::ALONG_ROWS,
+                           raft::sqrt_op{});
+      } else {
+        raft::linalg::norm(res,
+                           query,
+                           query_norms->view(),
+                           raft::linalg::NormType::L2Norm,
+                           raft::linalg::Apply::ALONG_ROWS);
+      }
+    }
+  }
+
+  /** a single batch of nearest neighbors in device memory */
+  class batch {
```

Comment:
This `batch` class could be re-used for batching IVF and CAGRA searches. Can it be implemented in a separate file?